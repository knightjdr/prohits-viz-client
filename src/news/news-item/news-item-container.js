import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import FetchNewsItem from '../../state/get/news-item-actions';
import NewsItem from './news-item';

export class NewsItemContainer extends Component {
  componentDidMount = () => {
    this.getNewsItemContent();
  }
  getNewsItemContent = () => {
    const { newsId } = this.props.match.params;
    if (newsId) {
      this.props.fetchNewsItem(newsId);
    }
  }
  render() {
    return (
      <NewsItem />
    );
  }
}

NewsItemContainer.defaultProps = {
  match: {
    params: {
      newsId: null,
    },
  },
};

NewsItemContainer.propTypes = {
  fetchNewsItem: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      newsId: PropTypes.string,
    }),
  }),
};

/* istanbul ignore next */

const mapDispatchToProps = dispatch => ({
  fetchNewsItem: (newsId) => {
    dispatch(FetchNewsItem(newsId));
  },
});

const ConnectedContainer = connect(
  null,
  mapDispatchToProps,
)(NewsItemContainer);

export default ConnectedContainer;
