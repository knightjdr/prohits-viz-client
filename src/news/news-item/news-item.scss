@import 'style/font';
@import 'style/button';
@import 'style/theme';

.news__item {
  max-width: 950px;
  width: calc(100vw - 50px);
}
.news__item-content {
  display: flex;
}
.news__item-date {
  font-size: 1.2rem;
}
.news__item-details {
  margin-top: 20px;
  text-align: justify;
}
.news__item-details a {
  text-decoration: underline;
}
.news__item-details a:link,
.news__item-details a:visited {
  color: $primary-medium;
}
.news__item-details a:focus,
.news__item-details a:hover,
.news__item-details a:active {
  color: $accent-light;
}
.news__item-headline {
  font-family: $TitleFont;
  font-size: 1.5rem;
  font-weight: bold;
}
.news__item-navlink,
.news__item-navlink:hover,
.news__item-navlink:visited, {
  color: $accent-dark;
}
.news__item-share-button {
  @extend .nobutton;
}
.news__item-share-buttons-title {
  font-weight: 600;
}
.news__item-navbutton-container {
  align-items: center;
  display: flex;
  flex-direction: row;
  left: 0;
  margin-top: 10px;
  position: relative;
  top: 0;
}
.news__item-navbutton-container > *:not(:last-child) {
  margin-right: 5px;
}
.news__item-share-buttons-title {
  margin-left: 10px;
}
.news__item-story {
  margin-left: 50px;
}
@media screen and (max-width: 679px) {
  .news__item-story {
    margin-left: 0;
  }
  .news__item-share-buttons-title {
    margin: 0 5px 0 10px;
  }
}
