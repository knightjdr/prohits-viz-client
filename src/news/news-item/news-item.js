/* eslint react/no-danger: 0 */

import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import PropTypes from 'prop-types';
import React from 'react';
import { Button, Tooltip } from 'antd';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';
import {
  EmailIcon,
  EmailShareButton,
  FacebookIcon,
  FacebookShareButton,
  TwitterIcon,
  TwitterShareButton,
} from 'react-share';
import { faNewspaper } from '@fortawesome/pro-regular-svg-icons';

import Loading from '../../components/loading/loading';
import NewsItemSelector from '../../state/selectors/news-item-selector';

import './news-item.css';

export const NewsItemComponent = ({
  newsItem,
}) => {
  let newsItemElement = null;
  if (newsItem.isLoading) {
    newsItemElement = (
      <Loading />
    );
  } else if (newsItem.error) {
    newsItemElement = (
      <Loading
        error
        message="There was an error retrieving this story"
      />
    );
  } else if (newsItem.isLoaded) {
    const newsButtons = (
      <div className="news__item-navbutton-container">
        <Tooltip
          placement="right"
          title="News archive"
        >
          <NavLink
            className="news__item-navlink"
            to="/news"
          >
            <Button
              type="primary"
              shape="circle"
            >
              <FontAwesomeIcon
                icon={faNewspaper}
              />
            </Button>
          </NavLink>
        </Tooltip>
        <div className="news__item-share-buttons-title">
          Share
        </div>
        <TwitterShareButton
          className="news__item-share-button"
          hashtags={newsItem.item.hashtags}
          title={newsItem.item.headline}
          url={`${process.env.REACT_APP_HOME_ROOT}/news/${encodeURI(newsItem.item.headline)}`}
        >
          <TwitterIcon size={32} round />
        </TwitterShareButton>
        <FacebookShareButton
          className="news__item-share-button"
          hashtag={newsItem.item.hashtags[0]}
          quote={newsItem.item.headline}
          url={`${process.env.REACT_APP_HOME_ROOT}/news/${encodeURI(newsItem.item.headline)}`}
        >
          <FacebookIcon size={32} round />
        </FacebookShareButton>
        <EmailShareButton
          className="news__item-share-button"
          subject={newsItem.item.headline}
          url={`${process.env.REACT_APP_HOME_ROOT}/news/${encodeURI(newsItem.item.headline)}`}
        >
          <EmailIcon size={32} round />
        </EmailShareButton>
      </div>
    );
    newsItemElement = (
      <div className="news__item-content">
        <article className="news__item-story">
          <header className="news__item-headline">
            { newsItem.item.headline }
          </header>
          <time className="news__item-date">
            { newsItem.item.date }
          </time>
          {newsButtons}
          <section
            className="news__item-details"
            dangerouslySetInnerHTML={{ __html: newsItem.item.html }}
          />
        </article>
      </div>
    );
  }
  return (
    <div className="news__item">
      { newsItemElement }
    </div>
  );
};

NewsItemComponent.defaultProps = {
  newsItem: null,
};

NewsItemComponent.propTypes = {
  newsItem: PropTypes.shape({
    error: PropTypes.bool,
    isLoaded: PropTypes.bool,
    isLoading: PropTypes.bool,
    item: PropTypes.shape({
      date: PropTypes.string,
      details: PropTypes.string,
      hashtags: PropTypes.arrayOf(
        PropTypes.string,
      ),
      headline: PropTypes.string,
      _id: PropTypes.string,
    }),
  }),
};

/* istanbul ignore next */

const mapStateToProps = state => ({
  newsItem: NewsItemSelector(state),
});

const ConnectedContainer = connect(
  mapStateToProps,
)(NewsItemComponent);

export default ConnectedContainer;
