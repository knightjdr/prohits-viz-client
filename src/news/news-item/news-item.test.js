import React from 'react';
import { shallow } from 'enzyme';

import textToHtml from '../../helpers/text-to-html';
import { NewsItemComponent } from './news-item';

// mock textToHtml
jest.mock('../../helpers/text-to-html');
textToHtml.mockReturnValue('test');

const testItem = {
  error: {
    error: true,
    isLoaded: false,
    isLoading: false,
    item: {},
  },
  init: {
    error: false,
    isLoaded: false,
    isLoading: false,
    item: {},
  },
  loaded: {
    error: false,
    isLoaded: true,
    isLoading: false,
    item: {
      date: 'test',
      details: 'test',
      hashtags: ['#test'],
      headline: 'test',
      _id: 'abcd',
    },
  },
  loading: {
    error: false,
    isLoaded: false,
    isLoading: true,
    item: {},
  },
};

describe('News item', () => {
  it('should render initially', () => {
    const wrapper = shallow(
      <NewsItemComponent
        newsItem={testItem.init}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('when loading', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = shallow(
        <NewsItemComponent
          newsItem={testItem.loading}
        />,
      );
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should show loading component', () => {
      expect(wrapper.find('Loading').length).toBe(1);
    });

    it('should have loading component with no error', () => {
      const loading = wrapper.find('Loading');
      expect(loading.props().error).toBeFalsy();
    });
  });

  describe('when loaded', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = shallow(
        <NewsItemComponent
          newsItem={testItem.loaded}
        />,
      );
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render new item', () => {
      expect(wrapper.find('.news__item-content').length).toBe(1);
    });
  });

  describe('with error', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = shallow(
        <NewsItemComponent
          newsItem={testItem.error}
        />,
      );
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should show loading component', () => {
      expect(wrapper.find('Loading').length).toBe(1);
    });

    it('should have loading component with error', () => {
      const loading = wrapper.find('Loading');
      expect(loading.props().error).toBeTruthy();
    });
  });
});
