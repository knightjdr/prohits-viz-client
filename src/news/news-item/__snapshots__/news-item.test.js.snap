// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`News item should render initially 1`] = `
<div
  className="news__item"
/>
`;

exports[`News item when loaded should match snapshot 1`] = `
<div
  className="news__item"
>
  <div
    className="news__item-content"
  >
    <article
      className="news__item-story"
    >
      <header
        className="news__item-headline"
      >
        test
      </header>
      <time
        className="news__item-date"
      >
        test
      </time>
      <div
        className="news__item-navbutton-container"
      >
        <Tooltip
          arrowPointAtCenter={false}
          autoAdjustOverflow={true}
          mouseEnterDelay={0.1}
          mouseLeaveDelay={0.1}
          placement="right"
          prefixCls="ant-tooltip"
          title="News archive"
          transitionName="zoom-big-fast"
        >
          <Component
            className="news__item-navlink"
            to="/news"
          >
            <Button
              ghost={false}
              loading={false}
              prefixCls="ant-btn"
              shape="circle"
              type="primary"
            >
              <FontAwesomeIcon$1
                border={false}
                className=""
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      576,
                      512,
                      Array [],
                      "f1ea",
                      "M552 64H112c-20.858 0-38.643 13.377-45.248 32H24c-13.255 0-24 10.745-24 24v272c0 30.928 25.072 56 56 56h496c13.255 0 24-10.745 24-24V88c0-13.255-10.745-24-24-24zM48 392V144h16v248c0 4.411-3.589 8-8 8s-8-3.589-8-8zm480 8H111.422c.374-2.614.578-5.283.578-8V112h416v288zM172 280h136c6.627 0 12-5.373 12-12v-96c0-6.627-5.373-12-12-12H172c-6.627 0-12 5.373-12 12v96c0 6.627 5.373 12 12 12zm28-80h80v40h-80v-40zm-40 140v-24c0-6.627 5.373-12 12-12h136c6.627 0 12 5.373 12 12v24c0 6.627-5.373 12-12 12H172c-6.627 0-12-5.373-12-12zm192 0v-24c0-6.627 5.373-12 12-12h104c6.627 0 12 5.373 12 12v24c0 6.627-5.373 12-12 12H364c-6.627 0-12-5.373-12-12zm0-144v-24c0-6.627 5.373-12 12-12h104c6.627 0 12 5.373 12 12v24c0 6.627-5.373 12-12 12H364c-6.627 0-12-5.373-12-12zm0 72v-24c0-6.627 5.373-12 12-12h104c6.627 0 12 5.373 12 12v24c0 6.627-5.373 12-12 12H364c-6.627 0-12-5.373-12-12z",
                    ],
                    "iconName": "newspaper",
                    "prefix": "far",
                  }
                }
                listItem={false}
                mask={null}
                name=""
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                symbol={false}
                transform={null}
              />
            </Button>
          </Component>
        </Tooltip>
        <div
          className="news__item-share-buttons-title"
        >
          Share
        </div>
        <CreatedButton
          className="news__item-share-button"
          hashtags={
            Array [
              "#test",
            ]
          }
          title="test"
          url="http://localhost:3000/news/test"
          windowHeight={400}
          windowWidth={550}
        >
          <Icon
            className=""
            iconBgStyle={Object {}}
            logoFillColor="white"
            round={true}
            size={32}
          />
        </CreatedButton>
        <CreatedButton
          className="news__item-share-button"
          hashtag="#test"
          quote="test"
          url="http://localhost:3000/news/test"
          windowHeight={400}
          windowWidth={550}
        >
          <Icon
            className=""
            iconBgStyle={Object {}}
            logoFillColor="white"
            round={true}
            size={32}
          />
        </CreatedButton>
        <CreatedButton
          className="news__item-share-button"
          onClick={[Function]}
          openWindow={false}
          subject="test"
          url="http://localhost:3000/news/test"
        >
          <Icon
            className=""
            iconBgStyle={Object {}}
            logoFillColor="white"
            round={true}
            size={32}
          />
        </CreatedButton>
      </div>
      <section
        className="news__item-details"
        dangerouslySetInnerHTML={
          Object {
            "__html": undefined,
          }
        }
      />
    </article>
  </div>
</div>
`;

exports[`News item when loading should match snapshot 1`] = `
<div
  className="news__item"
>
  <Loading
    error={false}
    message="There was an error loading this component"
    pastDelay={true}
  />
</div>
`;

exports[`News item with error should match snapshot 1`] = `
<div
  className="news__item"
>
  <Loading
    error={true}
    message="There was an error retrieving this story"
    pastDelay={true}
  />
</div>
`;
